tests/Unit/PackagePrivate/InMemoryTermIdsAcquirerTest.php

```diff
@@ -91,4 +92,78 @@ private function assertNoDuplicates( $ids ) {
 		);
 	}
 
+	public function testCleanTerms_doesNotReuseIds() {
+		$termIdsAcquirer = new InMemoryTermIdsAcquirer();
+
+		$ids1 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [
+				'en' => 'the label',
+				'de' => 'die Bezeichnung',
+			],
+			'alias' => [
+				'en' => [ 'alias', 'another' ],
+			],
+		] );
+		$ids2 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+
+		$termIdsAcquirer->cleanTerms( array_merge( $ids1, $ids2 ) );
+
+		$ids3 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+		$this->assertGreaterThan( max( max( $ids1 ), max( $ids2 ) ), min( $ids3 ) );
+	}
+
+	public function testCleanTerms_completelyCleansArray() {
+		$termIdsAcquirer = new InMemoryTermIdsAcquirer();
+
+		$ids1 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [
+				'en' => 'the label',
+				'de' => 'die Bezeichnung',
+			],
+			'alias' => [
+				'en' => [ 'alias', 'another' ],
+			],
+		] );
+		$ids2 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+
+		$termIdsAcquirer->cleanTerms( array_merge( $ids1, $ids2 ) );
+
+		$this->assertEmpty( TestingAccessWrapper::newFromObject( $termIdsAcquirer )->terms );
```

Comment:
I think breaking encapsulation like this is not a good idea. Since this was done deliberately I am not pushing the change onto this PR and instead made a follow up into it: https://github.com/wmde/mediawiki-term-store/pull/18

Reply:
Well, checking that `$terms` is completely empty (and not just something like `[ 'label' => [ 'en' => [] ]`) is the sole purpose of this test, and your follow-up no longer tests that. (`hasTerms()` might still return `false` for such a structure, depending on how it’s implemented.)

Reply:
Bit weird you are not putting that comment on the PR.

Will check later in my IDE myself - currently not clear to me how the behavior is changed. assertEmpty and empty() are not equivalent?

Reply:
As your pull request is a follow-up to this pull request, I think it makes sense to keep the discussion here – my comment was intended mainly as a reply to your comment, not as a review of your follow-up.

Neither your pull request nor mine use `empty()` as far as I can tell, but I assume your question refers to the implementation of `hasTerms()` in your pull request, which you then check with `assertFalse()`. The test `!== []` in that method is indeed equivalent to the `assertEmpty()` here (assuming the values are arrays), but this is an implementation detail – `hasTerms()` might be changed later so that it still returns `false` if the `$terms` for some reason end up in a state like `[ 'label' => [ 'en' => [] ] ]` (which still arguably has no actual terms). By directly using `assertEmpty` on the internal field in the tests, we avoid relying on this implementation detail.

Reply:
> By directly using assertEmpty on the internal field in the tests, we avoid relying on this implementation detail.

I am quite confused by this. Is it not the exact opposite? By using `hasTerms`, nothing outside of the class depends on its implementation, and by directly accessing the private field, the test does rely on it? Or are you talking about something else?

> As your pull request is a follow-up to this pull request, I think it makes sense to keep the discussion here – my comment was intended mainly as a reply to your comment, not as a review of your follow-up.

Not sure you noticed: my PRs are PRs *into* this one. If they are merged, they become part of this PR. Hence these are to be merged first, not after this PR, as you might expect when using the term "follow up".